File: src/runtime/policy.js
import { SEGMENT_ANON, SEGMENT_AUTH, SEGMENT_PUBLIC } from './key.js';

/**
 * Decides what may be cached, and for how long.
 *
 * Everything here is biased towards refusing to store. A missed caching
 * opportunity is a slow page; a wrong one serves a logged-in user's HTML to a
 * stranger. The rules below are the only thing standing between those.
 */

/** @param {string | null} value */
export function parse_cache_control(value) {
	const directives = new Map();
	if (!value) return directives;

	for (const part of value.split(',')) {
		const eq = part.indexOf('=');
		if (eq === -1) {
			directives.set(part.trim().toLowerCase(), true);
		} else {
			directives.set(part.slice(0, eq).trim().toLowerCase(), part.slice(eq + 1).trim());
		}
	}
	return directives;
}

function seconds(directives, name) {
	const raw = directives.get(name);
	if (raw === undefined || raw === true) return undefined;
	const n = Number.parseInt(String(raw).replace(/"/g, ''), 10);
	return Number.isFinite(n) && n >= 0 ? n : undefined;
}

/**
 * Can we even consider caching this request? Cheap checks, run before render.
 */
export function request_is_cacheable(request) {
	if (request.method !== 'GET' && request.method !== 'HEAD') return false;

	// A client that explicitly asked to skip the cache gets a fresh render, but
	// note this does NOT let a client poison the cache — we still store the
	// result normally, we just don't serve a stored one.
	const cc = parse_cache_control(request.headers.get('cache-control'));
	if (cc.has('no-cache') || cc.has('no-store')) return false;

	// An Authorization header means an API-style authenticated request. Those are
	// never cached under a shared segment and we don't have a segment for them.
	if (request.headers.has('authorization')) return false;

	return true;
}

/**
 * Decide whether to store the rendered response, in which segment, and with
 * what freshness. Returns null to refuse.
 *
 * @param {object} args
 * @param {Response} args.response
 * @param {boolean} args.authenticated  request carried an auth cookie
 * @param {object} args.ctx             per-request overrides from platform.cache
 * @param {object} args.config          resolved cache config
 */
export function decide_storage({ response, authenticated, ctx, config }) {
	// --- hard refusals, in order of how badly they'd bite -------------------

	// A route said no. Always wins.
	if (ctx.noStore) return null;

	// Only successful, complete responses. 206/304 have no full body to store,
	// and error pages should not be pinned for a TTL.
	if (response.status !== 200) return null;

	// Setting a cookie means the response is establishing or refreshing state
	// specific to this caller. Storing it would hand that state to whoever gets
	// the entry next. This catches the token-refresh path in hooks.server.ts
	// without the route needing to know anything about caching.
	if (response.headers.has('set-cookie')) return null;

	const cc = parse_cache_control(response.headers.get('cache-control'));
	if (cc.has('no-store') || cc.has('private')) return null;

	// Vary on anything we don't key on would mean storing one variant and
	// serving it to requests that should have received another. Cookie is called
	// out separately because it is the one that would leak rather than merely
	// mis-serve.
	const vary = (response.headers.get('vary') || '')
		.split(',')
		.map((v) => v.trim().toLowerCase())
		.filter(Boolean);

	if (vary.includes('*') || vary.includes('cookie') || vary.includes('authorization')) return null;

	const keyed = new Set([...config.varyHeaders.map((h) => h.toLowerCase()), 'accept-encoding']);
	for (const header of vary) {
		// accept-encoding is safe to ignore: we store bodies uncompressed and let
		// the proxy or Istio negotiate encoding on the way out.
		if (!keyed.has(header)) return null;
	}

	// --- segment ------------------------------------------------------------

	const segment = resolve_segment({ authenticated, ctx });
	if (!segment) return null;

	// --- freshness ----------------------------------------------------------

	const ttl = ctx.ttl ?? seconds(cc, 's-maxage') ?? config.defaultTtl;
	if (!ttl || ttl <= 0) return null;

	const swr = ctx.swr ?? seconds(cc, 'stale-while-revalidate') ?? config.defaultSwr;

	return {
		segment,
		ttl: Math.min(ttl, config.maxTtl),
		swr: Math.max(0, Math.min(swr ?? 0, config.maxTtl)),
		tags: [...ctx.tags]
	};
}

/**
 * Which segment may this response be stored in?
 *
 * The one rule that matters: a response rendered for a logged-in user never
 * lands in `pub` or `anon`. A route can declare itself public, but that
 * declaration is only honoured for requests that had no auth cookie — otherwise
 * a single mis-annotated route would leak on the first logged-in request.
 */
function resolve_segment({ authenticated, ctx }) {
	if (ctx.segment) {
		// Explicit per-user segment: isolated by construction, always allowed.
		if (ctx.segment.startsWith('user:')) return ctx.segment;

		if (ctx.segment === SEGMENT_AUTH) return authenticated ? SEGMENT_AUTH : null;

		if (ctx.segment === SEGMENT_PUBLIC) return authenticated ? null : SEGMENT_PUBLIC;
	}

	if (ctx.public) {
		// `public()` means "identical for everyone". Rendered anonymously it is
		// safe to share; rendered for a logged-in user we cannot verify the claim,
		// so we decline rather than trust it.
		return authenticated ? null : SEGMENT_PUBLIC;
	}

	// Default: an authed request is not cached at all. An anonymous one is cached
	// in its own segment, where no logged-in request will ever read it.
	return authenticated ? null : SEGMENT_ANON;
}

/**
 * May we serve this stored entry to this request?
 *
 * Belt-and-braces. Segment isolation already makes the wrong entry unreachable
 * by key, so this should never fire — but it is the check that turns a future
 * key-format bug into a cache miss instead of a data leak.
 */
export function may_serve(entry, { authenticated }) {
	if (!entry) return false;
	if (entry.segment === SEGMENT_PUBLIC) return true;
	if (entry.segment === SEGMENT_ANON) return !authenticated;
	if (entry.segment === SEGMENT_AUTH) return authenticated;
	if (entry.segment?.startsWith('user:')) return authenticated;
	return false;
}

/** Freshness of a stored entry, given the clock. */
export function freshness(entry, now) {
	const age = (now - entry.storedAt) / 1000;
	if (age <= entry.ttl) return 'fresh';
	if (age <= entry.ttl + entry.swr) return 'stale';
	return 'expired';
}

File: edge/policy.go
package main

import (
	"net/http"
	"strconv"
	"strings"
)

// Storage policy for responses from an arbitrary SSR origin.
//
// THIS FILE IS A CONTRACT, in the same way key.go is. It mirrors
// src/runtime/policy.js. When the SvelteKit adapter is in use it does its own
// storing and this path stays out of the way; for every other framework this is
// the only thing deciding what may be cached and who may see it.
//
// Everything here is biased towards refusing to store. A missed caching
// opportunity is a slow page; a wrong one serves a logged-in user's HTML to a
// stranger.

// Headers an origin uses to control caching without needing an adapter.
const (
	// Set by our own SvelteKit adapter so this proxy doesn't double-store what
	// the origin already wrote (with better information than we have).
	HeaderOrigin  = "X-Cache-Origin"
	HeaderTag     = "X-Cache-Tag"
	HeaderSegment = "X-Cache-Segment"
	HeaderTTL     = "X-Cache-TTL"
	HeaderSWR     = "X-Cache-SWR"
)

// Control headers are stripped before an entry is stored and before the
// response reaches the client — they are our transport, not the app's output.
var controlHeaders = []string{HeaderOrigin, HeaderTag, HeaderSegment, HeaderTTL, HeaderSWR}

type StorageDecision struct {
	Segment string
	TTL     int64
	SWR     int64
	Tags    []string
}

func parseCacheControl(value string) map[string]string {
	out := map[string]string{}
	if value == "" {
		return out
	}
	for _, part := range strings.Split(value, ",") {
		k, v, found := strings.Cut(part, "=")
		k = strings.ToLower(strings.TrimSpace(k))
		if k == "" {
			continue
		}
		if found {
			out[k] = strings.Trim(strings.TrimSpace(v), `"`)
		} else {
			out[k] = ""
		}
	}
	return out
}

func ccSeconds(cc map[string]string, name string) (int64, bool) {
	raw, ok := cc[name]
	if !ok || raw == "" {
		return 0, false
	}
	n, err := strconv.ParseInt(raw, 10, 64)
	if err != nil || n < 0 {
		return 0, false
	}
	return n, true
}

func headerSeconds(h http.Header, name string) (int64, bool) {
	raw := h.Get(name)
	if raw == "" {
		return 0, false
	}
	n, err := strconv.ParseInt(strings.TrimSpace(raw), 10, 64)
	if err != nil || n < 0 {
		return 0, false
	}
	return n, true
}

// RequestCacheable is the cheap pre-check, mirroring request_is_cacheable.
func RequestCacheable(r *http.Request) bool {
	if r.Method != http.MethodGet && r.Method != http.MethodHead {
		return false
	}
	if r.Header.Get("Authorization") != "" {
		return false
	}
	cc := strings.ToLower(r.Header.Get("Cache-Control"))
	if strings.Contains(cc, "no-cache") || strings.Contains(cc, "no-store") {
		return false
	}
	if r.Header.Get("Upgrade") != "" {
		return false
	}
	return true
}

// DecideStorage reports whether an upstream response may be stored, in which
// segment, and for how long. A nil return means refuse.
func DecideStorage(resp *http.Response, authenticated bool, cfg Config) *StorageDecision {
	// The SvelteKit adapter writes its own entries, with cache tags and segment
	// information this proxy cannot see. Storing again here would produce a
	// second, dumber copy under a key the adapter may consider stale.
	if resp.Header.Get(HeaderOrigin) != "" {
		return nil
	}

	// Only successful, complete responses. 206 and 304 have no full body to
	// store, and error pages must not be pinned for a TTL.
	if resp.StatusCode != http.StatusOK {
		return nil
	}

	// Setting a cookie means the response is establishing state specific to this
	// caller. Storing it would hand that state to whoever reads the entry next.
	if len(resp.Header.Values("Set-Cookie")) > 0 {
		return nil
	}

	cc := parseCacheControl(resp.Header.Get("Cache-Control"))
	if _, no := cc["no-store"]; no {
		return nil
	}
	if _, priv := cc["private"]; priv {
		return nil
	}

	if !varyIsKeyable(resp.Header, cfg) {
		return nil
	}

	segment := resolveSegment(resp.Header.Get(HeaderSegment), authenticated)
	if segment == "" {
		return nil
	}

	// Freshness must be explicit. An origin that says nothing gets nothing
	// cached — the same opt-in default as the adapter's defaultTtl of 0.
	ttl, ok := headerSeconds(resp.Header, HeaderTTL)
	if !ok {
		ttl, ok = ccSeconds(cc, "s-maxage")
	}
	if !ok || ttl <= 0 {
		return nil
	}
	if ttl > cfg.MaxTTL {
		ttl = cfg.MaxTTL
	}

	swr, ok := headerSeconds(resp.Header, HeaderSWR)
	if !ok {
		swr, _ = ccSeconds(cc, "stale-while-revalidate")
	}
	if swr > cfg.MaxTTL {
		swr = cfg.MaxTTL
	}

	return &StorageDecision{Segment: segment, TTL: ttl, SWR: swr, Tags: parseTags(resp.Header)}
}

// varyIsKeyable refuses any Vary we do not key on. Varying on something we
// ignore means storing one variant and serving it to requests that should have
// received another; Cookie and Authorization are called out because those would
// leak rather than merely mis-serve.
func varyIsKeyable(h http.Header, cfg Config) bool {
	keyed := map[string]bool{
		// Bodies are stored decoded, so encoding negotiation happens downstream.
		"accept-encoding": true,
	}
	for _, v := range cfg.VaryHeaders {
		keyed[strings.ToLower(strings.TrimSpace(v))] = true
	}

	for _, value := range h.Values("Vary") {
		for _, field := range strings.Split(value, ",") {
			f := strings.ToLower(strings.TrimSpace(field))
			if f == "" {
				continue
			}
			if f == "*" || f == "cookie" || f == "authorization" {
				return false
			}
			if !keyed[f] {
				return false
			}
		}
	}
	return true
}

// resolveSegment decides which segment a response may be stored in.
//
// The rule that matters: a response rendered for a logged-in user never lands
// in pub or anon. An origin can declare a page public, but that declaration is
// only honoured for requests that carried no auth cookie — otherwise a single
// mis-annotated route leaks on the first logged-in request.
func resolveSegment(declared string, authenticated bool) string {
	switch strings.ToLower(strings.TrimSpace(declared)) {
	case SegmentPublic:
		if authenticated {
			return ""
		}
		return SegmentPublic

	case SegmentAuth:
		if !authenticated {
			return ""
		}
		return SegmentAuth

	case SegmentAnon:
		if authenticated {
			return ""
		}
		return SegmentAnon
	}

	// Nothing declared. An authenticated request is not cached at all; an
	// anonymous one goes to its own segment, where no logged-in request can
	// reach it. Sharing with logged-in users requires opting in explicitly.
	if authenticated {
		return ""
	}
	return SegmentAnon
}

func parseTags(h http.Header) []string {
	var tags []string
	for _, value := range h.Values(HeaderTag) {
		for _, t := range strings.Split(value, ",") {
			if trimmed := strings.TrimSpace(t); trimmed != "" {
				tags = append(tags, trimmed)
			}
		}
	}
	return tags
}

// storableHeaders copies response headers minus the ones that must not be
// replayed to the next visitor.
func storableHeaders(h http.Header) map[string]string {
	skip := map[string]bool{
		"set-cookie":        true,
		"content-encoding":  true,
		"content-length":    true,
		"transfer-encoding": true,
		"connection":        true,
		"keep-alive":        true,
		"upgrade":           true,
	}
	for _, c := range controlHeaders {
		skip[strings.ToLower(c)] = true
	}

	out := map[string]string{}
	for k, v := range h {
		if skip[strings.ToLower(k)] || len(v) == 0 {
			continue
		}
		out[k] = v[0]
	}
	return out
}

func stripControlHeaders(h http.Header) {
	for _, c := range controlHeaders {
		h.Del(c)
	}
}
